wgengine/magicsock/endpoint.go

```diff
@@ -354,6 +359,44 @@ func (de *endpoint) addrForWireGuardSendLocked(now mono.Time) (udpAddr netip.Add
 	return udpAddr, needPing
 }
 
+// addrForPingSizeLocked returns the address(es) that should be used for sending
+// the next ping. It will only return addrs with a large enough path MTU to
+// permit a ping payload of size bytes to be delivered (DERP is always one such
+// addr as it is a TCP connection). If it returns a zero-value udpAddr, then we
+// should continue probing the MTU of all paths to this endpoint. Zero, one, or
+// both of the returned UDP address and DERP address may be non-zero.
+//
+// de.mu must be held.
+func (de *endpoint) addrForPingSizeLocked(now mono.Time, size int) (udpAddr, derpAddr netip.AddrPort) {
+	if size == 0 {
+		udpAddr, derpAddr, _ = de.addrForSendLocked(now)
+		return
+	}
+
+	udpAddr = de.bestAddr.AddrPort
+	mtuOk := true
+	pathMTU := de.bestAddr.wireMTU
+	requestedMTU := pingSizeToPktLen(size, udpAddr.Addr())
+	if requestedMTU > pathMTU {
+		mtuOk = false
+	}
+
+	if udpAddr.IsValid() && mtuOk {
+		if !now.After(de.trustBestAddrUntil) {
+			return udpAddr, netip.AddrPort{}
+		}
+		// We had a bestAddr with large enough MTU but it expired, so
+		// send both to it and DERP.
+		return udpAddr, de.derpAddr
+	}
+
+	// The UDP address isn't valid or it doesn't have a path MTU big enough
+	// for the packet. Return a zero-value udpAddr to signal that we should
+	// keep probing the path MTU to all addresses for this endpoint, and a
+	// valid DERP addr to signal that we should also send via DERP.
```

Comment:
Thank you for detailed comments explaining what's going on here! 👍

Reply:
I am excited about understanding this well enough to eventually rewrite the addrFor*Locked() in ways that don't require detailed comments to understand them. :)